File: Packages/NMP/Sources/NMP/Nip46SessionCheckpoint.swift
import Foundation
import NMPFFI

/// Distinguishes a NIP-46 session paired via `nostrconnect://` from one
/// dialed via `bunker://` (#571). Restore mechanics are identical either
/// way -- kept because "absence of a reusable client checkpoint is
/// observable rather than guessed from partial metadata" is a hard
/// requirement, not because restore branches on it.
public enum NMPNip46SessionOrigin: Sendable, Equatable {
    case clientInitiated
    case bunker

    init(_ ffi: FfiNip46Origin) {
        switch ffi {
        case .clientInitiated: self = .clientInitiated
        case .bunker: self = .bunker
        }
    }

    func toFfi() -> FfiNip46Origin {
        switch self {
        case .clientInitiated: .clientInitiated
        case .bunker: .bunker
        }
    }
}

/// NMP-owned checkpoint of the minimum secrets and descriptor needed to
/// reconnect an already-authorized NIP-46 client session without another
/// pairing handshake (#571) -- read from `NMPNip46Connection.checkpoint()`
/// once `.ready`, persisted by an `NMPNip46SessionCheckpointStore`, and
/// consumed by `NMPEngine.restoreNip46Session`/`importNip46Session`.
///
/// `clientSecretKey` hands the raw secret to whoever holds this value, by
/// the same design as `NMPLocalAccountCheckpoint` (see that protocol's
/// doc): the app owns its keys, and the recommended secure checkpoint
/// stores keep it in the platform vault. Nothing here ever logs, prints, or
/// otherwise surfaces it -- `description`/`debugDescription` are redacted.
public struct NMPNip46SessionCheckpoint: Sendable, Equatable {
    public let clientSecretKey: String
    public let userPublicKey: String
    public let remoteSignerPublicKey: String
    public let relays: [String]
    public let origin: NMPNip46SessionOrigin

    public init(
        clientSecretKey: String,
        userPublicKey: String,
        remoteSignerPublicKey: String,
        relays: [String],
        origin: NMPNip46SessionOrigin
    ) {
        self.clientSecretKey = clientSecretKey
        self.userPublicKey = userPublicKey
        self.remoteSignerPublicKey = remoteSignerPublicKey
        self.relays = relays
        self.origin = origin
    }

    init(_ ffi: FfiNip46SessionCheckpoint) {
        clientSecretKey = ffi.clientSecretKey
        userPublicKey = ffi.userPublicKey
        remoteSignerPublicKey = ffi.remoteSignerPublicKey
        relays = ffi.relays
        origin = NMPNip46SessionOrigin(ffi.origin)
    }

    func toFfi() -> FfiNip46SessionCheckpoint {
        FfiNip46SessionCheckpoint(
            clientSecretKey: clientSecretKey,
            userPublicKey: userPublicKey,
            remoteSignerPublicKey: remoteSignerPublicKey,
            relays: relays,
            origin: origin.toFfi()
        )
    }
}

extension NMPNip46SessionCheckpoint: CustomStringConvertible, CustomDebugStringConvertible {
    public var description: String {
        "NMPNip46SessionCheckpoint(userPublicKey: \(userPublicKey), " +
            "remoteSignerPublicKey: \(remoteSignerPublicKey), relays: \(relays), " +
            "origin: \(origin), clientSecretKey: [redacted])"
    }

    public var debugDescription: String { description }
}

/// NMP-owned versioned wire serialization of `NMPNip46SessionCheckpoint`,
/// independent of the FFI-generated type's own field order/renames (#571).
extension NMPNip46SessionCheckpoint {
    private struct WireV1: Codable {
        var version: Int
        var clientSecretKey: String
        var userPublicKey: String
        var remoteSignerPublicKey: String
        var relays: [String]
        var origin: String
    }

    /// A serialized checkpoint blob did not decode as this format -- never
    /// carries the raw bytes, so this is always safe to log.
    public enum SerializationError: Error, Equatable {
        case unsupportedVersion(Int)
        case invalidOrigin(String)
        case malformed
    }

    /// Encode this checkpoint as NMP's own versioned wire format. A
    /// checkpoint store persists this `Data` verbatim.
    public func serialize() throws -> Data {
        let wire = WireV1(
            version: 1,
            clientSecretKey: clientSecretKey,
            userPublicKey: userPublicKey,
            remoteSignerPublicKey: remoteSignerPublicKey,
            relays: relays,
            origin: origin == .clientInitiated ? "clientInitiated" : "bunker"
        )
        return try JSONEncoder().encode(wire)
    }

    /// Decode a checkpoint previously produced by `serialize()`. An
    /// unrecognized version, corrupt bytes, or an unknown origin tag fails
    /// closed with a typed `SerializationError` rather than guessing.
    public static func deserialize(_ data: Data) throws -> NMPNip46SessionCheckpoint {
        let wire: WireV1
        do {
            wire = try JSONDecoder().decode(WireV1.self, from: data)
        } catch {
            throw SerializationError.malformed
        }
        guard wire.version == 1 else {
            throw SerializationError.unsupportedVersion(wire.version)
        }
        let origin: NMPNip46SessionOrigin
        switch wire.origin {
        case "clientInitiated": origin = .clientInitiated
        case "bunker": origin = .bunker
        default: throw SerializationError.invalidOrigin(wire.origin)
        }
        return NMPNip46SessionCheckpoint(
            clientSecretKey: wire.clientSecretKey,
            userPublicKey: wire.userPublicKey,
            remoteSignerPublicKey: wire.remoteSignerPublicKey,
            relays: wire.relays,
            origin: origin
        )
    }
}

/// The seam an `NMPEngine` NIP-46 restore/import door reads/writes through,
/// mirroring `NMPLocalAccountCheckpoint`'s load/save/clear shape (#571): a
/// platform-vault provider, an app-custom store, or an app-supplied
/// convenience type are all drop-ins. A checkpoint hands the raw client
/// secret to its holder by the same design as `NMPLocalAccountCheckpoint`
/// (see that protocol's doc) -- the app owns its keys.
public protocol NMPNip46SessionCheckpointStore: Sendable {
    /// The checkpointed session, or `nil` when none is stored.
    func loadCheckpoint() throws -> NMPNip46SessionCheckpoint?
    /// Persist `checkpoint`, replacing any previous one.
    func saveCheckpoint(_ checkpoint: NMPNip46SessionCheckpoint) throws
    /// Remove the checkpoint. Idempotent -- a later `loadCheckpoint` returns
    /// `nil`. Never called by `close()`/`deinit`; only explicit session
    /// removal clears it.
    func clear() throws
}

/// The raw generated FFI record (`NMPFFI.FfiNip46SessionCheckpoint`) has no
/// redacted `Debug` of its own -- unlike `nmp-signer`'s Rust type it
/// mirrors, a plain Swift struct's default reflection-based description
/// prints every stored property, including `clientSecretKey`. This
/// retroactive conformance closes that gap for the one boundary moment the
/// raw FFI value exists (inside `NMPNip46SessionCheckpoint.init(_:)` and
/// `NMPNip46Connection.checkpoint()`) before it is wrapped into the redacted
/// `NMPNip46SessionCheckpoint` apps actually hold.
extension FfiNip46SessionCheckpoint: CustomStringConvertible, CustomDebugStringConvertible {
    public var description: String {
        "FfiNip46SessionCheckpoint(userPublicKey: \(userPublicKey), " +
            "remoteSignerPublicKey: \(remoteSignerPublicKey), relays: \(relays), " +
            "origin: \(origin), clientSecretKey: [redacted])"
    }

    public var debugDescription: String { description }
}

File: Packages/NMPKotlin/src/main/kotlin/com/nmp/sdk/Nip46SessionCheckpoint.kt
package com.nmp.sdk

import java.nio.charset.StandardCharsets
import uniffi.nmp_ffi.FfiNip46Origin
import uniffi.nmp_ffi.FfiNip46SessionCheckpoint

/** Distinguishes a NIP-46 session paired via `nostrconnect://` from one
 * dialed via `bunker://` (#571). Restore mechanics are identical either way
 * -- kept because "absence of a reusable client checkpoint is observable
 * rather than guessed from partial metadata" is a hard requirement, not
 * because restore branches on it. */
enum class NMPNip46SessionOrigin {
    ClientInitiated,
    Bunker,
    ;

    internal fun toFfi(): FfiNip46Origin =
        when (this) {
            ClientInitiated -> FfiNip46Origin.CLIENT_INITIATED
            Bunker -> FfiNip46Origin.BUNKER
        }

    internal companion object {
        fun from(ffi: FfiNip46Origin): NMPNip46SessionOrigin =
            when (ffi) {
                FfiNip46Origin.CLIENT_INITIATED -> ClientInitiated
                FfiNip46Origin.BUNKER -> Bunker
            }
    }
}

/**
 * NMP-owned checkpoint of the minimum secrets and descriptor needed to
 * reconnect an already-authorized NIP-46 client session without another
 * pairing handshake (#571) -- read from [NMPNip46Connection.checkpoint]
 * once `Ready`, persisted by an [NMPNip46SessionCheckpointStore], and
 * consumed by [restoreNip46Session]/[importNip46Session].
 *
 * [clientSecretKey] hands the raw secret to whoever holds this value, by
 * the same design as [NMPLocalAccountCheckpoint] (see that interface's
 * doc): the app owns its keys, and the recommended secure checkpoint
 * stores keep it in a secure store. [toString] is redacted -- it never
 * prints [clientSecretKey].
 */
data class NMPNip46SessionCheckpoint(
    val clientSecretKey: String,
    val userPublicKey: String,
    val remoteSignerPublicKey: String,
    val relays: List<String>,
    val origin: NMPNip46SessionOrigin,
) {
    internal constructor(ffi: FfiNip46SessionCheckpoint) : this(
        clientSecretKey = ffi.clientSecretKey,
        userPublicKey = ffi.userPublicKey,
        remoteSignerPublicKey = ffi.remoteSignerPublicKey,
        relays = ffi.relays,
        origin = NMPNip46SessionOrigin.from(ffi.origin),
    )

    internal fun toFfi(): FfiNip46SessionCheckpoint =
        FfiNip46SessionCheckpoint(
            clientSecretKey = clientSecretKey,
            userPublicKey = userPublicKey,
            remoteSignerPublicKey = remoteSignerPublicKey,
            relays = relays,
            origin = origin.toFfi(),
        )

    /** Redacted -- never includes [clientSecretKey]. */
    override fun toString(): String =
        "NMPNip46SessionCheckpoint(userPublicKey=$userPublicKey, " +
            "remoteSignerPublicKey=$remoteSignerPublicKey, relays=$relays, " +
            "origin=$origin, clientSecretKey=[redacted])"

    /**
     * Encode this checkpoint as NMP's own versioned wire format: one
     * newline-delimited record (version tag, then every field, relays
     * length-prefixed) -- independent of the FFI-generated type's own field
     * order/renames, and dependency-free (no JSON library is a main-scope
     * dependency of this module). A checkpoint store persists these bytes
     * verbatim.
     */
    fun serialize(): ByteArray {
        val lines = mutableListOf(WIRE_VERSION.toString(), origin.name, clientSecretKey, userPublicKey, remoteSignerPublicKey, relays.size.toString())
        lines.addAll(relays)
        return lines.joinToString("\n").toByteArray(StandardCharsets.UTF_8)
    }

    /**
     * A serialized checkpoint blob did not decode as this format.
     *
     * This wire format is POSITIONAL (fixed line indices), unlike Swift's
     * keyed-JSON equivalent: a blob missing or shifted by one line can make
     * an unrelated field's raw content -- including the client secret --
     * land at the exact line index this decoder reads as "version" or
     * "origin". Every message below is therefore purely positional/
     * descriptive (line index, counts) and NEVER interpolates the actual
     * line content read from the blob, so this is always safe to log even
     * against a corrupt or adversarially shifted blob.
     */
    sealed class SerializationException(message: String) : Exception(message) {
        class UnsupportedVersion(val version: Int) :
            SerializationException("unsupported NIP-46 checkpoint wire version: $version")

        object InvalidOrigin :
            SerializationException("checkpoint wire line 1 (origin) did not match a known origin tag")

        class Malformed(reason: String) : SerializationException("malformed NIP-46 checkpoint: $reason")
    }

    companion object {
        private const val WIRE_VERSION = 1

        /** Decode a checkpoint previously produced by [serialize]. An
         * unrecognized version, corrupt bytes, or an unknown origin tag
         * fails closed with a typed [SerializationException] rather than
         * guessing. */
        fun deserialize(data: ByteArray): NMPNip46SessionCheckpoint {
            val lines = String(data, StandardCharsets.UTF_8).split("\n")
            if (lines.size < 6) {
                throw SerializationException.Malformed("expected at least 6 lines, got ${lines.size}")
            }
            val version = lines[0].toIntOrNull()
                ?: throw SerializationException.Malformed("wire line 0 (version) was not numeric")
            if (version != WIRE_VERSION) {
                throw SerializationException.UnsupportedVersion(version)
            }
            val origin = when (lines[1]) {
                NMPNip46SessionOrigin.ClientInitiated.name -> NMPNip46SessionOrigin.ClientInitiated
                NMPNip46SessionOrigin.Bunker.name -> NMPNip46SessionOrigin.Bunker
                else -> throw SerializationException.InvalidOrigin
            }
            val relayCount = lines[5].toIntOrNull()
                ?: throw SerializationException.Malformed("wire line 5 (relay count) was not numeric")
            val relays = lines.drop(6)
            if (relays.size != relayCount) {
                throw SerializationException.Malformed(
                    "declared $relayCount relays but found ${relays.size}",
                )
            }
            return NMPNip46SessionCheckpoint(
                clientSecretKey = lines[2],
                userPublicKey = lines[3],
                remoteSignerPublicKey = lines[4],
                relays = relays,
                origin = origin,
            )
        }
    }
}

/**
 * The seam an [NMPEngine] NIP-46 restore/import door reads/writes through,
 * mirroring [NMPLocalAccountCheckpoint]'s load/save/clear shape (#571): a
 * secure-store provider, an app-custom store, or an app-supplied
 * convenience type are all drop-ins. A checkpoint hands the raw client
 * secret to its holder by the same design as [NMPLocalAccountCheckpoint]
 * (see that interface's doc) -- the app owns its keys.
 */
interface NMPNip46SessionCheckpointStore {
    /** The checkpointed session, or `null` when none is stored. */
    fun loadCheckpoint(): NMPNip46SessionCheckpoint?

    /** Persist [checkpoint], replacing any previous one. */
    fun saveCheckpoint(checkpoint: NMPNip46SessionCheckpoint)

    /** Remove the checkpoint. Idempotent -- a later [loadCheckpoint] returns
     * `null`. Never called by [NMPNip46Connection.close]; only explicit
     * session removal clears it. */
    fun clear()
}
